File: backend/transactions/ledger_grouping.py
"""Group flat ledger transaction dicts by calendar year, then month (descending)."""

import calendar
from collections import OrderedDict
from typing import Any


def group_ledger_rows_by_year_month(rows: list[dict[str, Any]]) -> list[dict[str, Any]]:
    """
    Rows must be ordered newest-first; each row must include ``txn_date`` (``YYYY-MM-DD``).

    Returns::
        [
            {"year": 2026, "months": [
                {"month": 4, "label": "April 2026", "transactions": [...]},
                ...
            ]},
            ...
        ]
    Years descending; months descending within each year.
    """
    years: OrderedDict[int, OrderedDict[int, list]] = OrderedDict()
    for row in rows:
        td = str(row.get("txn_date") or "")
        parts = td.split("-")
        if len(parts) < 2:
            continue
        try:
            y = int(parts[0])
            m = int(parts[1])
        except ValueError:
            continue
        if y not in years:
            years[y] = OrderedDict()
        if m not in years[y]:
            years[y][m] = []
        years[y][m].append(row)

    out: list[dict[str, Any]] = []
    for y in sorted(years.keys(), reverse=True):
        months_block: list[dict[str, Any]] = []
        for m in sorted(years[y].keys(), reverse=True):
            label = f"{calendar.month_name[int(m)]} {y}"
            months_block.append(
                {
                    "month": int(m),
                    "label": label,
                    "transactions": years[y][m],
                }
            )
        out.append({"year": y, "months": months_block})
    return out

File: web/src/utils/ledgerGrouping.ts
export type LedgerMonthGroup<T> = {
  month: number;
  label: string;
  transactions: T[];
};

export type LedgerYearGroup<T> = {
  year: number;
  months: LedgerMonthGroup<T>[];
};

const MONTH_NAMES = [
  "January",
  "February",
  "March",
  "April",
  "May",
  "June",
  "July",
  "August",
  "September",
  "October",
  "November",
  "December",
];

export function groupLedgerRowsByYearMonth<T extends { txn_date: string }>(
  rows: T[]
): LedgerYearGroup<T>[] {
  const years = new Map<number, Map<number, T[]>>();

  for (const row of rows) {
    const parts = String(row.txn_date ?? "").split("-");
    if (parts.length < 2) continue;
    const y = parseInt(parts[0], 10);
    const m = parseInt(parts[1], 10);
    if (!Number.isFinite(y) || !Number.isFinite(m)) continue;

    if (!years.has(y)) years.set(y, new Map());
    const ym = years.get(y)!;
    if (!ym.has(m)) ym.set(m, []);
    ym.get(m)!.push(row);
  }

  const yearOrder = Array.from(years.keys()).sort((a, b) => b - a);
  const out: LedgerYearGroup<T>[] = [];
  for (const y of yearOrder) {
    const ym = years.get(y)!;
    const months = Array.from(ym.keys()).sort((a, b) => b - a);
    const monthsBlock: LedgerMonthGroup<T>[] = [];
    for (const m of months) {
      const txs = ym.get(m);
      if (!txs?.length) continue;
      monthsBlock.push({
        month: m,
        label: `${MONTH_NAMES[m - 1] ?? m} ${y}`,
        transactions: txs,
      });
    }
    if (monthsBlock.length) out.push({ year: y, months: monthsBlock });
  }
  return out;
}

